content/courses/Ada_For_The_C_Embedded_Developer/chapters/10_Conclusion.rst

```diff
@@ -3,6 +3,175 @@ Conclusion
+equivalent. One reason is that the two languages are fairly similar in the way
+they implement imperative semantics, in particular with regards to memory
+management or control flow. Therefore, they should be equivalent on average.
+However, when a piece of code in Ada is significantly slower than its
+counterpart in C, this usually comes from the fact that, while the two pieces
+of code appear to be semantically equivalent, they happen to be actually quite
+different. Fortunately, there are strategies that we can use to improve the
+performance and make it equivalent to the C version. These are some examples:
+
+- Clever use of compilation switches, which might optimize the performance of
+  an application significantly.
+
+- Suppression of checks at specific parts of the implementation.
+
+    - Although runtime checks are very useful and should be used as much as
+      possible, they can also increase the overhead of implementations at
+      certain hot-spots.
+
+- Restriction of assertions to development code.
+
+    - For example, we may use assertions in the debug version of the code and
+      turn them off in the release version.
+
+    - Also, we may use formal proof to decide which assertions we turn off in
+      the release version. By formally proving that assertions will never fail
+      at run-time, we can safely deactivate them.
+
+Formal proof |mdash| a form of static analysis |mdash| can give strong
+guarantees about checks, for all possible conditions and all possible inputs.
+It verifies conditions prior to execution, even prior to compilation, so we can
+remove bugs earlier in the development phase. This is far less expensive than
+doing so later because the cost to fix bugs increases exponentially over the
+phases of the project life cycle, especially after deployment. Preventing bug
+introduction into the deployed system is the least expensive approach of all.
+
+Formal analysis for proof can be achieved through the SPARK subset of the Ada
+language combined with the :program:`gnatprove` verification tool. SPARK is a
+subset encompassing most of the Ada language, except for features that preclude
+proof.
+
+In Ada, several common programming errors that are not already detected at
+compile-time are detected instead at run-time, triggering *exceptions* that
+interrupt the normal flow of execution. However, we may be able to prove that
+the language-defined checks won't raise exceptions at run-time. This is known
+as proving *Absence of Run-Time Errors*. Successful proof of these checks
+is highly significant in itself. One of the major resulting benefits is that we
+can deploy the final executable with checks disabled.
+
+In many situations, the migration of C code to Ada is justified by an increase
+in terms of integrity expectations, in which case it's expected that
+development costs will raise. However, because Ada reduces the level of
+processes, expertise and tools needed in a project, it makes it possible to
```

Comment:
I don't think Ada reduces the level of processes, expertise, and tools, leading to reduced costs. I think it is a more expressive, powerful language, designed to reduce errors earlier in the life-cycle, thus reducing costs. My 2 cents....

Reply:
I like your suggestion — I'm using it in the new version of the PR.